prov/sockets: Allow multiple threads to wait on one counter

jpdoyle opened the pull request:
fi_cntr_wait cannot be called on a single counter from multiple threads, even with the FI_WAIT_MUTEX_COND wait object (it returns -FI_EBUSY). This patch allows that use case.

lanl-ompi commented:
bot:retest

j-xiong commented:
The negation of error code is unnecessary and doesn't make the behavior consistent within the file. If you want to change, please change sock_cntr_open the other way around instead. That will be consistent with the remaining of the code.

jpdoyle commented:
I was not aware of which way was the current style in the codebase -- I have changed the negation.

jithinjosepkl commented:
@jpdoyle - Thanks for the patch. It will be good to allow fi_cntr_wait() be called from multiple threads.

jithinjosepkl commented on prov/sockets/src/sock_cntr.c:
```diff
@@ -210,8 +210,9 @@ void sock_cntr_inc(struct sock_cntr *cntr)
 	atomic_inc(&cntr->value);
-	if (atomic_get(&cntr->value) >= atomic_get(&cntr->threshold))
-		pthread_cond_signal(&cntr->cond);
+	/* if(atomic_get(&cntr->waiting)) { */
```
Is there a reason for commenting this condition? Since this is in the critical path, it makes sense to call the signal() only if somebody is waiting on the counter. Also, I had noticed some performance penalties if we call signal() always.

jpdoyle replied:
I believe I misread the context around this line. I was worried about a race between reading cntr->waiting and signalling the condition variable, but now I see we already have the mutex here.

jithinjosepkl commented on prov/sockets/src/sock_cntr.c:
```diff
@@ -233,8 +234,9 @@ static int sock_cntr_add(struct fid_cntr *cntr, uint64_t value)
-	if (atomic_get(&_cntr->value) >= atomic_get(&_cntr->threshold))
-		pthread_cond_signal(&_cntr->cond);
+	/* if(atomic_get(&cntr->waiting)) { */
+	pthread_cond_broadcast(&_cntr->cond);
```
same comment as above

jithinjosepkl commented on prov/sockets/src/sock_cntr.c:
```diff
@@ -274,13 +277,13 @@ static int sock_cntr_wait(struct fid_cntr *cntr, uint64_t threshold,
 
-	if (_cntr->is_waiting) {
+	if (atomic_get(&_cntr->num_waiting) &&
+			_cntr->domain->progress_mode == FI_PROGRESS_MANUAL) {
```
Can't we support this even for manual-progress mode?

jpdoyle replied:
Is `sock_cntr_progress` thread safe? It looks like it is at a glance, but I'm not familiar enough with the surrounding code-base to make that judgement.

jithinjosepkl replied:
Yes, it is.

jithinjosepkl commented on prov/sockets/src/sock_cntr.c:
```diff
@@ -292,20 +295,36 @@ static int sock_cntr_wait(struct fid_cntr *cntr, uint64_t threshold,
+		remaining_ms = timeout;
+		if (timeout >= 0) {
+			start_ms = fi_gettime_ms();
+			end_ms = start_ms + timeout;
```
We are doing in the if - true case. We can just move this up before the if statement.

jithinjosepkl commented on prov/sockets/src/sock_poll.c:
```diff
@@ -141,13 +141,19 @@ static int sock_poll_poll(struct fid_poll *pollset, void **context, int count)
+						old_val = new_val;
+						new_val = atomic_cswap(&cntr->old_value, old_val, val);
+					} while(new_val != old_val && new_val < val);
+				}
```
This part might not be needed as part of this patch. I think fi_poll() with counters need some clarification (please see https://github.com/ofiwg/libfabric/issues/2240). May be we can drop this (along with the atomic_cswap implementation) from the PR.

jpdoyle replied:
The change this patch makes to the sock_cntr structure (removing threshold) invalidates the existing fi_poll implementation. We could leave threshold in the struct and have a dead code path (since no threshold would ever be set), or return -FI_EINVAL in this branch.

This part of the patch implements a reasonable (I think) interpretation of fi_poll's semantics by returning any counters whose values have changed since the last fi_poll call -- consistent with the CQ and EQ branches returning whenever there is new data available. If it is better to leave dead code or return -FI_EINVAL, I can update the patch accordingly.

jithinjosepkl replied:
I think the behavior of fi_poll() with counters needs some clarification. Till then, I think it is better to return -FI_EINVAL. Once the usage is clarified(or updated with new interface), we can implement it.

shantonu commented:
Noticed some minor formatting issues. If we break one long line to multiple ones then the next line is aligned by tab and then spaces till two spaces after the opening parenthesis of the first line.

jpdoyle commented:
I've removed the sock_poll changes, fixed a few formatting issues, allowed multi-threaded sock_cntr_wait in manual-progress mode, and squashed the commits.

jithinjosepkl commented:
👍

j-xiong commented on prov/sockets/src/sock_cntr.c:
```diff
@@ -210,8 +210,12 @@ void sock_cntr_inc(struct sock_cntr *cntr)
 	atomic_inc(&cntr->value);
-	if (atomic_get(&cntr->value) >= atomic_get(&cntr->threshold))
-		pthread_cond_signal(&cntr->cond);
+	/* Note: This will not deadlock sock_cntr_wait because we
```
This comment is not necessary.

j-xiong replied:
In addition, no need to put username and date in comment.

j-xiong commented on prov/sockets/src/sock_cntr.c:
```diff
@@ -210,8 +210,12 @@ void sock_cntr_inc(struct sock_cntr *cntr)
+	/* Note: This will not deadlock sock_cntr_wait because we
+	 * already have the mutex. -jpdoyle 05-Aug-2016
+	 */
+	if(atomic_get(&cntr->num_waiting)) {
```
Remove the braces.

j-xiong commented on prov/sockets/src/sock_cntr.c:
```diff
@@ -233,8 +237,9 @@ static int sock_cntr_add(struct fid_cntr *cntr, uint64_t value)
 	atomic_set(&_cntr->value, atomic_get(&_cntr->value) + value);
-	if (atomic_get(&_cntr->value) >= atomic_get(&_cntr->threshold))
-		pthread_cond_signal(&_cntr->cond);
+	if(atomic_get(&_cntr->num_waiting)) {
```
Remove the braces.

j-xiong commented on prov/sockets/src/sock_cntr.c:
```diff
@@ -247,18 +252,19 @@ static int sock_cntr_set(struct fid_cntr *cntr, uint64_t value)
 	atomic_set(&_cntr->value, value);
-	if (atomic_get(&_cntr->value) >= atomic_get(&_cntr->threshold))
-		pthread_cond_signal(&_cntr->cond);
+	if(atomic_get(&_cntr->num_waiting)) {
```
Remove the braces.

j-xiong commented on prov/sockets/src/sock_poll.c:
```diff
@@ -63,10 +63,6 @@ static int sock_poll_add(struct fid_poll *pollset, struct fid *event_fid,
 		cq = container_of(list_item->fid, struct sock_cq, cq_fid);
 		atomic_inc(&cq->ref);
 		break;
-	case FI_CLASS_CNTR:
```
Mention this change in the commit message.

jpdoyle commented:
I've rebased onto the new master with the updates to fi_poll and @j-xiong's formatting suggestions. I also combined the loops for automatic and manual progress so that their shared semantics are clearer. I noticed that if `sock_cntr_progress` returns an error code, it would not be returned by `sock_cntr_wait`. Currently, `sock_cntr_progress` only ever returns 0, but that seems like a possible bug -- `sock_progress_tx_ctx` and `sock_progress_rx_ctx` can return errors.
